packages/financial-templates-lib/src/clients/OptimisticOracleClient.js

```diff
@@ -73,54 +100,118 @@ class OptimisticOracleClient {
+    // Store proposals that have NOT been disputed:
+    let undisputedProposals = proposalEvents.filter(event => {
+      const key = this._getPriceRequestKey(event);
+      const hasDispute = disputeEvents.find(disputeEvent => this._getPriceRequestKey(disputeEvent) === key);
+      return hasDispute === undefined;
+    });
+    undisputedProposals = undisputedProposals.map(event => {
+      return {
+        requester: event.returnValues.requester,
+        proposer: event.returnValues.proposer,
+        identifier: this.hexToUtf8(event.returnValues.identifier),
+        timestamp: event.returnValues.timestamp,
+        proposedPrice: event.returnValues.proposedPrice,
+        expirationTimestamp: event.returnValues.expirationTimestamp
       };
-      if (state === OptimisticOracleRequestStatesEnum.REQUESTED) {
-        this.priceRequests.push(requestData);
-      } else if (state === OptimisticOracleRequestStatesEnum.PROPOSED) {
-        // TODO: Add data specific for proposals,
-        // or alternatively search for ProposedPrice events.
-        this.priceProposals.push(requestData);
-      }
     });
 
+    // Filter proposals based on their expiration timestamp:
+    const isExpired = proposal => {
+      return Number(proposal.expirationTimestamp) <= Number(currentTime);
+    };
+    this.expiredProposals = undisputedProposals.filter(proposal => isExpired(proposal));
+    this.undisputedProposals = undisputedProposals.filter(proposal => !isExpired(proposal));
+
+    // Store disputes that can be settled:
+    let resolvedPrices = await Promise.all(
+      disputeEvents.map(async disputeEvent => {
+        try {
+          // When someone disputes an OO proposal, the OO requests a price to the DVM with a re-formatted
+          // ancillary data packet that includes the original requester's information:
+          const stampedAncillaryData = await this.oracle.methods
+            .stampAncillaryData(
+              disputeEvent.returnValues.ancillaryData ? disputeEvent.returnValues.ancillaryData : "0x",
+              disputeEvent.returnValues.requester
+            )
+            .call();
+
+          // getPrice will return null or revert if there is no price available,
+          // in which case we'll ignore this dispute.
+          let resolvedPrice = revertWrapper(
+            await this.voting.methods
+              .getPrice(disputeEvent.returnValues.identifier, disputeEvent.returnValues.timestamp, stampedAncillaryData)
+              .call({
+                from: this.oracle.options.address
+              })
+          );
+          if (resolvedPrice !== null) {
+            return disputeEvent;
+          }
+        } catch (error) {
+          // No resolved price available, do nothing.
+        }
+      })
+    );
+    // Remove undefined entries, marking disputes that did not have resolved prices
+    resolvedPrices = resolvedPrices.filter(event => event !== undefined);
```

Comment:
nit: not strongly opinionated (and this may be less readable), but you could chain this with the last line so you don't have to reassign this variable?

Reply:
Yes I think this is more readable actually.